lib/index.js

```diff
@@ -103,410 +35,425 @@ const defaultRules = [
-
-        // otherwise, follow ignore-walk's logic
-        : super.filterEntry(entry, partial)
-    )
+  // overridden method: we use a custom sort method to help compressibility
+  sort (a, b) {
+    // optimize for compressibility
+    // extname, then basename, then locale alphabetically
+    // https://twitter.com/isntitvacant/status/1131094910923231232
+    const exta = extname(a).toLowerCase()
+    const extb = extname(b).toLowerCase()
+    const basea = basename(a).toLowerCase()
+    const baseb = basename(b).toLowerCase()
+
+    return exta.localeCompare(extb, 'en') ||
+      basea.localeCompare(baseb, 'en') ||
+      a.localeCompare(b, 'en')
   }
 
-  filterEntries () {
-    if (this.ignoreRules['.npmignore']) {
-      this.ignoreRules['.gitignore'] = null
-    }
-    this.filterEntries = super.filterEntries
-    super.filterEntries()
+  // convenience method: this joins the given rules with newlines, appends a trailing newline,
+  // and calls the internal onReadIgnoreFile method
+  injectRules (filename, rules, callback = () => {}) {
+    this.onReadIgnoreFile(filename, `${rules.join('\n')}\n`, callback)
   }
 
-  addIgnoreFile (file, then) {
-    const ig = path.resolve(this.path, file)
-    if (file === 'package.json' && !this.isProject) {
-      then()
-    } else if (this.packageJsonCache.has(ig)) {
-      this.onPackageJson(ig, this.packageJsonCache.get(ig), then)
-    } else {
-      super.addIgnoreFile(file, then)
+  // custom method: this is called by addIgnoreFile when we find a package.json, it uses the
+  // arborist tree to pull both default rules and strict rules for the package
+  processPackage (callback) {
+    const {
+      bin,
+      browser,
+      files,
+      main,
+    } = this.tree.package
+
+    // rules in these arrays are inverted since they are patterns we want to _not_ ignore
+    const ignores = []
+    const strict = [
+      ...strictDefaults,
+      '!/package.json',
+      '!/npm-shrinkwrap.json',
+      '/.git',
+      '/node_modules',
+      '/package-lock.json',
+      '/yarn.lock',
+      '/pnpm-lock.yaml',
+    ]
+
+    // if we have a files array in our package, we need to pull rules from it
+    if (files) {
+      for (const file of files) {
+        // invert the rule because these are things we want to include
+        const inverse = `!${file}`
+        try {
+          // if an entry in the files array is a specific file, then we need to include it as a
+          // strict requirement for this package. if it's a directory or a pattern, it's a default
+          // pattern instead. this is ugly, but we have to stat to find out if it's a file
+          const stat = lstat(join(this.path, file.replace(/^!+/, '')).replace(/\\/g, '/'))
+          // if we have a file and we know that, it's strictly required
+          if (stat.isFile()) {
+            strict.unshift(inverse)
+            this.requiredFiles.push(file.startsWith('/') ? file.slice(1) : file)
```

Comment:
is this faster than:
```suggestion
            this.requiredFiles.push(file.replace(/^\//, ''))
```
?

Reply:
essentially the same. this is more for our general guiding principle of avoiding regular expressions